src/Telemetry/AzureDevOpsTimeline/AzureDevOpsTimeline.cs

```diff
@@ -319,5 +364,56 @@ private static string GetBucket(AugmentedTimelineIssue augIssue)
         {
             return await azureServer.ListBuilds(project, cancellationToken, minDateTime, limit);
         }
+
+        private async Task GetImageNames(List<AugmentedTimelineRecord> records, CancellationToken cancellationToken)
+        {
+            SemaphoreSlim throttleSemaphore = new SemaphoreSlim(50);
+
+            var bag = new ConcurrentBag<Task>();
+            int canceledTasksNumber = 0;
+
+            foreach(var record in records)
+            {
+                if(!string.IsNullOrEmpty(record.Raw.Log?.Url) && record.Raw.Name == "Initialize job")
+                {
+                    var childTask = Task.Run(async () =>
+                    {
+                        try
+                        {
+                            await throttleSemaphore.WaitAsync();
+
+                            if (!cancellationToken.IsCancellationRequested)
+                            {
+                                record.ImageName = (record.Raw.WorkerName.StartsWith("Azure Pipelines") || record.Raw.WorkerName.StartsWith("Hosted Agent"))
+                                ? await _buildLogScraper.ExtractMicrosoftHostedPoolImageNameAsync(record.Raw.Log.Url, cancellationToken)
+                                : record.Raw.WorkerName.StartsWith("NetCore1ESPool-")
+                                    ? await _buildLogScraper.ExtractOneESHostedPoolImageNameAsync(record.Raw.Log.Url, cancellationToken)
+                                    : null;
+                            }
+                            else
+                            {
+                                Interlocked.Increment(ref canceledTasksNumber);
+                            }
+                        }
+                        catch (OperationCanceledException) 
+                        {
+                            Interlocked.Increment(ref canceledTasksNumber);
+                        }
+                        catch (Exception exception)
+                        {
+                            _logger.LogInformation($"Non critical exception thrown when trying to get log '{record.Raw.Log.Url}': {exception}");
+                        }
+                        finally
+                        {
+                            throttleSemaphore.Release();
+                        }
+                    }, cancellationToken);
+                    bag.Add(childTask);
+                }
+            }
+
+            await Task.WhenAll(bag.ToArray());
+            _logger.LogInformation($"Log scraping had {canceledTasksNumber} out of {bag.Count} tasks canceled due to timeout");
```

Comment:
I _think_ you could possibly just analyze the `bag` and whether tasks completed, and you can skip a whole lot of code for cancelled task counting.

Also, alternatively, you could have an increment counter at the end for success and count success, which should be easier and get the number from there

Reply:
We originally tried using the bag to get number of cancelled tasks, but it is always 0, as we handle all exceptions in the task itself, so it always finishes properly.
I think counting cancelled tasks is better that counting successful ones, since the number is lower in most cases, and we don't have to do any subtraction

Reply:
Don't swallow stuff inside the tasks themselves and then you solve your first problem.

Reply:
The problem with not swallowing stuff inside the task, is that we're catching the exception from the parent thread, exiting the function early. Is there a way to avoid sending exceptions from child threads to parent threads?

I'll rewrite the code to count successes

Reply:
You can catch the "AggregateException" from the Task.WhenAll and examine the task list (or just count how many "inner exceptions" are in the aggregate).  No need for fancy multithreaded counting stuff.  I'm also not sure having that information in the log is really "worth" all of the added complexity. "Any" failing is probably enough of a signal (the difference between, say 4 failing and 7 failing isn't really significant)

Reply:
AggregateException was catching this because OperationCancelledException is thrown at least one task is cancelled. I refactored it a bit